python/ray/serve/utils.py

```diff
@@ -9,37 +9,70 @@
 from typing import List
 import io
 import os
+from ray.serve.exceptions import RayServeException
 
 import requests
+import pydantic
+from werkzeug.datastructures import ImmutableMultiDict
 
 import ray
 from ray.serve.constants import HTTP_PROXY_TIMEOUT
-from ray.serve.context import FakeFlaskRequest, TaskContext
+from ray.serve.context import TaskContext
 from ray.serve.http_util import build_flask_request
 import numpy as np
 
-try:
-    import pydantic
-except ImportError:
-    pydantic = None
-
 ACTOR_FAILURE_RETRY_TIMEOUT_S = 60
 
 
+class ServeRequest:
+    def __init__(self, data, kwargs, headers, method):
+        self._data = data
+        self._kwargs = kwargs
+        self._headers = headers
+        self._method = method
+
+    @property
+    def method(self):
+        return self._method
+
+    @property
+    def args(self):
+        return ImmutableMultiDict(self._kwargs)
+
+    @property
+    def json(self):
+        if not isinstance(self._data, dict):
+            raise RayServeException("Request data is not a dictionary. "
+                                    f"It is {type(self._data)}.")
+        return self._data
+
+    @property
+    def form(self):
+        if not isinstance(self._data, dict):
+            raise RayServeException("Request data is not a dictionary. "
+                                    f"It is {type(self._data)}.")
+        return self._data
+
+    @property
+    def data(self):
+        return self._data
```

Comment:
In the dict case should we lazily serialize this to json?

Reply:
That's probably what I'd expect as a user

Reply:
I think we shouldn't do it. Because the only time you will be accessing the http_request.data is when the data is not json encoded. If the data is json encoded, you can just call `http_request.json` to get it. This means in http_request.data case, it won't be serialized json.

We should keep ServeRequest.data to return whatever Python object it's passed in. This allow users to pass in, say Torch Tensor, Image file, etc.

Reply:
Sounds fine to me. It's easier to explain